### PR TITLE
RDBC-1053: Implemented several fixes.

### PR DESCRIPTION
**Pull request by redknightlois**
### Issue link

https://issues.hibernatingrhinos.com/issue/RDBC-1053

### Additional description

  - **BatchPatchCommandData**: duplicate document IDs silently accepted instead of throwing
  - **AbstractDocumentQuery.closeSubclause**: nested subclause boost iteration doesn't stop after finding the matching open token
  - **JsonOperation**: field type change from scalar to array produces spurious array change records alongside the correct FieldChanged
  - **AbstractDocumentQuery range queries**: `whereBetween`, `whereGreaterThan`, `whereLessThan`, `whereGreaterThanOrEqual`, `whereLessThanOrEqual` all treat `0`, `""`, and `false` as
  null/unbounded
  - **AbstractDocumentQuery.isFilterActive**: checks bottom of filter mode stack instead of top, breaking nested filter scopes
  - **DocumentConventions.identityPartsSeparator**: setter validates the current value instead of the incoming value, allowing `"|"` on first set and blocking all subsequent changes
  - **TimeValue.compareTo**: cross-unit comparison (months vs seconds) compares the same value's bounds against itself instead of the other value's bounds
  - **TimeValue._isSpecialCompare**: `MIN_VALUE.compareTo(MIN_VALUE)` returns `-1` instead of `0` due to swapped min/max checks
  - **AbstractDocumentQuery._fuzzy/_proximity**: null token list causes `TypeError` instead of the intended error message

### Type of change

- [x] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [ ] New feature

### How risky is the change?

- [x] Low
- [ ] Moderate
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [ ] No documentation update is needed

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`)
- [ ] It has been verified by manual testing
- [ ] Existing tests verify the correct behavior

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

### CODE REVIEW

File: src/Documents/Session/AbstractDocumentQuery.ts [hunk 1]
```diff
@@ -213,7 +213,7 @@ export abstract class AbstractDocumentQuery<T extends object, TSelf extends Abst
     protected _explanationToken: ExplanationToken;
 
     protected isFilterActive(): boolean {
-        return this._filterModeStack.length && this._filterModeStack[0];
+        return this._filterModeStack.length && this._filterModeStack[this._filterModeStack.length - 1];
```
Comment (M4xymm) [inline src/Documents/Session/AbstractDocumentQuery.ts]: This snippet can be written shorter, but it does not change the result.

```suggestion
        return this._filterModeStack.length && this._filterModeStack.at(-1);
```

File: test/Documents/Session/AbstractDocumentQueryTest.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    afterEach(() => {
+        store?.dispose();
+    });
+
+    function createStore() {
+        store = new DocumentStore("http://localhost:9999", "TestDB");
+        store.conventions.findCollectionNameForObjectLiteral = () => "test";
+        store.initialize();
+        return store;
+    }
```
Comment (M4xymm) [inline test/Documents/Session/AbstractDocumentQueryTest.ts]: example of testing with DocumentStore

https://github.com/ravendb/ravendb-nodejs-client/blob/c5d5e9805f94e2c829dd1268da163a7db8942141/test/Ported/Documents/Operations/GenAiEtlTest.ts#L15-L22

```suggestion
    let store: IDocumentStore;

    beforeEach(async function () {
        store = await testContext.getDocumentStore();
    });

    afterEach(async () =>
        await disposeTestDocumentStore(store));
```

File: test/Documents/Session/AbstractDocumentQueryTest.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    it("whereBetween preserves 0 as a valid boundary value", function () {
+        const s = createStore().openSession();
+
+        const q = s.query({ collection: "users" }).whereBetween("age", 0, 100);
+        const iq = q.getIndexQuery();
+
+        assert.strictEqual(iq.queryParameters.p0, 0,
+            "start value 0 should be preserved, not replaced with wildcard '*'");
+        assert.strictEqual(iq.queryParameters.p1, 100);
+    });
```
Comment (M4xymm) [inline test/Documents/Session/AbstractDocumentQueryTest.ts:26]: I know these are just tests, but it would be better to avoid abbreviations like `q` or `iq`, as they result in poor naming. Variable names should be descriptive enough that you can understand their purpose at a glance :)

```suggestion
    it("whereBetween preserves 0 as a valid boundary value", function () {
        const query = store.query({ collection: "users" }).whereBetween("age", 0, 100);
        const indexQuery = query.getIndexQuery();

        assert.strictEqual(indexQuery.queryParameters.p0, 0,
            "start value 0 should be preserved, not replaced with wildcard '*'");
        assert.strictEqual(indexQuery.queryParameters.p1, 100);
    });
```

**Review by M4xymm — CHANGES_REQUESTED** [review]

File: test/Mapping/JsonOperationTest.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import assert from "node:assert";
+import { describe, it } from "node:test";
```
Comment (redknightlois) [inline test/Mapping/JsonOperationTest.ts:2]: @M4xymm this is because I cannot make it work with older mocha... maybe we should upgrade first. https://issues.hibernatingrhinos.com/issue/RDBC-1055/Upgrade-mocha-from-10.7.3-to-11.x

**Review by M4xymm — APPROVED** [review]
